pkg/controllers/nodeclaim/disruption/consolidation.go

```diff
@@ -0,0 +1,86 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package disruption
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/samber/lo"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/metrics"
+)
+
+// Consolidation is a nodeclaim sub-controller that adds or removes status conditions on empty nodeclaims based on consolidateAfter
+type Consolidation struct {
+	kubeClient client.Client
+	clock      clock.Clock
+}
+
+//nolint:gocyclo
+func (c *Consolidation) Reconcile(ctx context.Context, nodePool *v1.NodePool, nodeClaim *v1.NodeClaim) (reconcile.Result, error) {
+	hasConsolidatableCondition := nodeClaim.StatusConditions().Get(v1.ConditionTypeConsolidatable) != nil
+
+	// 1. If Consolidation isn't enabled, remove the consolidatable status condition
+	if nodePool.Spec.Disruption.ConsolidateAfter.Duration == nil {
+		if hasConsolidatableCondition {
+			_ = nodeClaim.StatusConditions().Clear(v1.ConditionTypeConsolidatable)
+			log.FromContext(ctx).V(1).Info("removing consolidatable status condition, consolidation is disabled")
+		}
+		return reconcile.Result{}, nil
+	}
+	initialized := nodeClaim.StatusConditions().Get(v1.ConditionTypeInitialized)
+	// 2. If NodeClaim is not initialized, remove the consolidatable status condition
+	if !initialized.IsTrue() {
+		if hasConsolidatableCondition {
+			_ = nodeClaim.StatusConditions().Clear(v1.ConditionTypeConsolidatable)
+			log.FromContext(ctx).V(1).Info("removing consolidatable status condition, isn't initialized")
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// If the lastPodEvent is zero, use the time that the nodeclaim was initialized, as that's when Karpenter recognizes that pods could have started scheduling
+	timeToCheck := lo.Ternary(!nodeClaim.Status.LastPodEventTime.IsZero(), nodeClaim.Status.LastPodEventTime.Time, initialized.LastTransitionTime.Time)
+
+	// Consider a node consolidatable by looking at the lastPodEvent status field on the nodeclaim.
+	if c.clock.Since(timeToCheck) < lo.FromPtr(nodePool.Spec.Disruption.ConsolidateAfter.Duration) {
+		if hasConsolidatableCondition {
+			_ = nodeClaim.StatusConditions().Clear(v1.ConditionTypeConsolidatable)
+			log.FromContext(ctx).V(1).Info("removing consolidatable status condition")
+		}
+		consolidatableTime := nodeClaim.Status.LastPodEventTime.Add(lo.FromPtr(nodePool.Spec.Disruption.ConsolidateAfter.Duration))
+		return reconcile.Result{RequeueAfter: time.Until(consolidatableTime)}, nil
```

Comment:
`time.Until` mixes the clock with the global `time.Time` IIRC. So, you probably don't want to use this function because you are going to be mixed `time.Now()` with the time value that you actually control through the clock interface